drivers/serial/uart_numaker.c

```diff
@@ -0,0 +1,484 @@
+{
+	switch (parity) {
+	case UART_CFG_PARITY_ODD:
+		return UART_PARITY_ODD;
+	case UART_CFG_PARITY_EVEN:
+		return UART_PARITY_EVEN;
+	case UART_CFG_PARITY_MARK:
+		return UART_PARITY_MARK;
+	case UART_CFG_PARITY_SPACE:
+		return UART_PARITY_SPACE;
+	case UART_CFG_PARITY_NONE:
+	default:
+		return UART_PARITY_NONE;
+	}
+}
+
+#ifdef CONFIG_UART_USE_RUNTIME_CONFIGURE
+static int uart_numaker_configure(const struct device *dev, const struct uart_config *cfg)
+{
+	const struct uart_numaker_config *config = dev->config;
+	struct uart_numaker_data *pData = dev->data;
+	int32_t databits, stopbits;
+	uint32_t parity;
+
+	databits = uart_numaker_convert_datalen(cfg->data_bits);
+	if (databits < 0) {
+		return databits;
+	}
+
+	stopbits = uart_numaker_convert_stopbit(cfg->stop_bits);
+	if (stopbits < 0) {
+		return stopbits;
+	}
+
+	if (cfg->flow_ctrl == UART_CFG_FLOW_CTRL_NONE) {
+		UART_DisableFlowCtrl(config->uart);
+	} else if (cfg->flow_ctrl == UART_CFG_FLOW_CTRL_RTS_CTS) {
+		UART_EnableFlowCtrl(config->uart);
+	} else {
+		return -ENOTSUP;
+	}
+
+	parity = uart_numaker_convert_parity(cfg->parity);
+
+	UART_SetLineConfig(config->uart, cfg->baudrate, databits, parity, stopbits);
+
+	memcpy(&pData->ucfg, cfg, sizeof(*cfg));
+
+	return 0;
+}
+
+static int uart_numaker_config_get(const struct device *dev, struct uart_config *cfg)
+{
+	struct uart_numaker_data *pData = dev->data;
+
+	memcpy(cfg, &pData->ucfg, sizeof(*cfg));
+
+	return 0;
+}
+#endif /* CONFIG_UART_USE_RUNTIME_CONFIGURE */
+
+static int uart_numaker_init(const struct device *dev)
+{
+	const struct uart_numaker_config *config = dev->config;
+	struct uart_numaker_data *pData = dev->data;
+	int err = 0;
+
+	SYS_UnlockReg();
+
+#ifdef CONFIG_CLOCK_CONTROL_NUMAKER_SCC
+	struct numaker_scc_subsys scc_subsys;
+
+	memset(&scc_subsys, 0x00, sizeof(scc_subsys));
+	scc_subsys.subsys_id = NUMAKER_SCC_SUBSYS_ID_PCC;
+	scc_subsys.pcc.clk_modidx = config->clk_modidx;
+	scc_subsys.pcc.clk_src = config->clk_src;
+	scc_subsys.pcc.clk_div = config->clk_div;
+
+	/* Equivalent to CLK_EnableModuleClock() */
+	err = clock_control_on(config->clk_dev, (clock_control_subsys_t)&scc_subsys);
+	if (err != 0) {
+		goto move_exit;
+	}
+	/* Equivalent to CLK_SetModuleClock() */
+	err = clock_control_configure(config->clk_dev, (clock_control_subsys_t)&scc_subsys, NULL);
+	if (err != 0) {
+		goto move_exit;
+	}
+#else
+	/* Enable UART module clock */
+	CLK_EnableModuleClock(config->clk_modidx);
+
+	/* Select UART module clock source/divider */
+	CLK_SetModuleClock(config->clk_modidx, config->clk_src, config->clk_div);
+#endif
+
+	/* Set pinctrl for UART0 RXD and TXD */
+	/* Set multi-function pins for UART0 RXD and TXD */
+#ifdef CONFIG_PINCTRL
+	err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (err != 0) {
+		goto move_exit;
+	}
+#else
+	/* For M46x EVB debug port */
+	SET_UART0_RXD_PB12();
+	SET_UART0_TXD_PB13();
+#endif
```

Comment:
Could the `#else` part be done via board config in dts? You can mandate that this driver requires `PINCTRL` so it can always set the pins. I am worried about the time when developers would disable `PINCTRL` manually for debugging and the else part is executed.

Reply:
Thanks of your consideration.
The `#else` part is one specific case and a little bit confused.
So, I removed it in commit [7295a30](https://github.com/zephyrproject-rtos/zephyr/pull/55406/commits/7295a308aaa8ce08cbb207a4d2ae7234503f88b6).

Reply:
@dcpleung , 
Hi Daniel,
I rebase to fix CI failure of `"\tests\bsim folder not foud"`.
So, the relative commit changed as [91b193c](https://github.com/zephyrproject-rtos/zephyr/pull/55406/commits/91b193cbc3d373b16a83257de0483536d6bf1dce)